File: digdag-standards/src/main/resources/digdag/standards/rb/runner.rb
require 'json'

command = ARGV[0]
out_file = ARGV[2]

module DigdagEnv
  in_file = ARGV[1]

  in_data = JSON.parse(File.read(in_file))
  params = in_data['params']

  # TODO include indifferent access like Embulk::DataSource
  PARAMS = params
  SUBTASK_CONFIG = {}
  EXPORT_PARAMS = {}
  STORE_PARAMS = {}
  STATE_PARAMS = {}
end

# should this be a digdag.gem so that users can unit-test a command without running digdag?
module Digdag
  class Env
    def initialize
      @params = DigdagEnv::PARAMS
      @subtask_config = DigdagEnv::SUBTASK_CONFIG
      @export_params = DigdagEnv::EXPORT_PARAMS
      @store_params = DigdagEnv::STORE_PARAMS
      @state_params = DigdagEnv::STATE_PARAMS
      @subtask_index = 0
    end

    attr_reader :params

    attr_reader :subtask_config

    attr_reader :export_params

    attr_reader :store_params

    attr_reader :state_params

    def set_state(**params)
      @state_params.merge!(params)
    end

    def export(**params)
      @export_params.merge!(params)
    end

    def store(**params)
      @store_params.merge!(params)
    end

    # add_subtask(params)
    # add_subtask(singleton_method_name, params={})
    # add_subtask(klass, instance_method_name, params={})
    def add_subtask(*args)
      if args.length == 1 && args[0].is_a?(Hash)
        # add_subtask(params)
        config = args[0]

      elsif args.length == 1 || (args.length == 2 && args[1].is_a?(Hash))
        # add_subtask(singleton_method_name, params={})
        method_name = args[0]
        params = Hash(args[1])

        begin
          method_name = method_name.to_sym
        rescue NameError, ArgumentError
          raise ArgumentError, "Second argument must be a Symbol but got #{method_name.inspect}"
        end

        if method_name.to_s.include?(".")
          raise ArgumentError, "Method name can't include '.'"
        end

        config = params.dup
        config["rb>"] = method_name.to_s

      elsif args.length == 2 || (args.length == 3 && args[2].is_a?(Hash))
        # add_subtask(klass, instance_method_name, params={})
        klass = args[0]
        method_name = args[1]
        params = Hash(args[2])

        begin
          method_name = method_name.to_sym
        rescue NameError, ArgumentError
          raise ArgumentError, "Second argument must be a Symbol but got #{method_name.inspect}"
        end

        if method_name.to_s.include?(".")
          raise ArgumentError, "Method name can't include '.'"
        end

        if klass.is_a?(Class)
          class_name = klass.name

        else
          begin
            class_name = klass.to_sym.to_s
          rescue NameError, ArgumentError
            raise ArgumentError, "First argument must be a Class or Symbol but got #{klass.inspect}"
          end
        end

        # validation
        begin
          klass = Kernel.const_get(class_name)  # const_get with String (not Symbol) searches nested constants
        rescue NameError
          raise ArgumentError, "Could not find class named #{class_name}"
        end

        unless klass.respond_to?(method_name) || klass.public_instance_methods.include?(method_name)
          raise ArgumentError, "Class #{klass} does not have method #{method_name.inspect}"
        end

        config = params.dup
        config["rb>"] = "::#{klass}.#{method_name}"

      else
        raise ArgumentError, "wrong number of arguments (#{args.length} for 1..3 with the last argument is a Hash)"
      end

      begin
        JSON.dump(config)
      rescue => e
        raise ArgumentError, "Parameters must be serializable using JSON: #{e}"
      end

      @subtask_config["+subtask#{@subtask_index}"] = config
      @subtask_index += 1

      nil
    end
  end

  DIGDAG_ENV = Env.new
  private_constant :DIGDAG_ENV

  def self.env
    DIGDAG_ENV
  end
end

# add the archive path to LOAD_PATH
$LOAD_PATH << File.expand_path(Dir.pwd)

def digdag_inspect_command(command)
  fragments = command.split(".")
  method_name = fragments.pop.to_sym
  if fragments.empty?
    # method
    return nil, method_name, false
  else
    # Name::Space::Class.method
    class_name = fragments.join(".")
    klass = Kernel.const_get(class_name)
    is_instance_method = klass.public_instance_methods.include?(method_name)
    return klass, method_name, is_instance_method
  end
end

def digdag_inspect_arguments(receiver, method_name, params)
  if receiver
    parameters = receiver.method(method_name).parameters
    if method_name == :new && parameters == [[:rest]]
      # This is Object.new that forwards all arguments to #initialize
      begin
        parameters = receiver.instance_method(:initialize).parameters
      rescue NameError => e
      end
    end
  else
    parameters = method(method_name).parameters
  end

  args = []
  keywords = nil
  parameters.each do |kind,name|
    key = name.to_s
    case kind
    when :req
      # required argument like a
      unless params.has_key?(key)
        if receiver.is_a?(Class)
          raise ArgumentError, "Method '#{receiver}.#{method_name}' requires parameter '#{key}' but not set"
        else
          raise ArgumentError, "Method '#{receiver.class}##{method_name}' requires parameter '#{key}' but not set"
        end
      end
      args << params[key]

    when :opt
      # optional argument like a=nil
      if params.has_key?(key)
        args << params[key]
      else
        # use the default value.
      end

    when :rest
      # variable-length arguments like *a
      # there're really we can do here to keep consistency with :opt.
      # should this be an error?

    when :keyreq
      # required keyword argument like a:
      unless params.has_key?(key)
        if receiver.is_a?(Class)
          raise ArgumentError, "Method '#{receiver}.#{method_name}' requires parameter '#{key}' but not set"
        else
          raise ArgumentError, "Method '#{receiver.class}##{method_name}' requires parameter '#{key}' but not set"
        end
      end
      if keywords.nil?
        keywords = {}
        args << keywords
      end
      keywords[name] = params[key]

    when :key
      # optional keyword argument like a: nil
      if params.has_key?(key)
        if keywords.nil?
          keywords = {}
          args << keywords
        end
        keywords[name] = params[key]
      else
        # use the default value.
      end

    when :keyrest
      # rest-of-keywords argument like **a
      # symbolize keys otherwise method call causes error:
      # "TypeError: wrong argument type String (expected Symbol)"
      if keywords.nil?
        keywords = {}
        args << keywords
      end
      keywords.merge!(digdag_symbolize_keys(params))
    end
  end

  return args
end

def digdag_symbolize_keys(hash)
  built = {}
  hash.each_pair do |k,v|
    if v.is_a?(Hash)
      v = digdag_symbolize_keys(v)
    end
    built[k.to_s.to_sym] = v
  end
  return built
end

klass, method_name, is_instance_method = digdag_inspect_command(command)
error = nil

if klass.nil?
  method_args = digdag_inspect_arguments(nil, method_name, DigdagEnv::PARAMS)
  begin
    result = send(method_name, *method_args)
  rescue => error
  end

elsif is_instance_method
  new_args = digdag_inspect_arguments(klass, :new, DigdagEnv::PARAMS)
  instance = klass.new(*new_args)

  method_args = digdag_inspect_arguments(instance, method_name, DigdagEnv::PARAMS)
  begin
    result = instance.send(method_name, *method_args)
  rescue => error
  end

else
  method_args = digdag_inspect_arguments(klass, method_name, DigdagEnv::PARAMS)
  begin
    result = klass.send(method_name, *method_args)
  rescue => error
  end
end

out = {
  'subtask_config' => DigdagEnv::SUBTASK_CONFIG,
  'export_params' => DigdagEnv::EXPORT_PARAMS,
  'store_params' => DigdagEnv::STORE_PARAMS,
  #'state_params' => DigdagEnv::STATE_PARAMS,  # only for retrying
}

if error
  out['error'] = {
    'class' => error.class.to_s,
    'message' => error.message,
    'backtrace' => error.backtrace,
  }
end

File.open(out_file, "w") {|f| f.write out.to_json }

raise error if error

File: digdag-standards/src/main/resources/digdag/standards/py/runner.py
import sys
import os
import json
import types
import inspect
import collections
import traceback

command = sys.argv[1]
in_file = sys.argv[2]
out_file = sys.argv[3]

with open(in_file) as f:
    in_data = json.load(f)
    params = in_data['params']

# fake digdag_env module already imported
digdag_env_mod = sys.modules['digdag_env'] = types.ModuleType('digdag_env')
digdag_env_mod.params = params
digdag_env_mod.subtask_config = collections.OrderedDict()
digdag_env_mod.export_params = {}
digdag_env_mod.store_params = {}
digdag_env_mod.state_params = {}
import digdag_env

# fake digdag module already imported
digdag_mod = sys.modules['digdag'] = types.ModuleType('digdag')

class Env(object):
    def __init__(self, digdag_env_mod):
        self.params = digdag_env_mod.params
        self.subtask_config = digdag_env_mod.subtask_config
        self.export_params = digdag_env_mod.export_params
        self.store_params = digdag_env_mod.store_params
        self.state_params = digdag_env_mod.state_params
        self.subtask_index = 0

    def set_state(self, params={}, **kwds):
        self.state_params.update(params)
        self.state_params.update(kwds)

    def export(self, params={}, **kwds):
        self.export_params.update(params)
        self.export_params.update(kwds)

    def store(self, params={}, **kwds):
        self.store_params.update(params)
        self.store_params.update(kwds)

    def add_subtask(self, function=None, **params):
        if function is not None and not isinstance(function, dict):
            if hasattr(function, "im_class"):
                # Python 2
                command = ".".join([function.im_class.__module__, function.im_class.__name__, function.__name__])
            else:
                # Python 3
                command = ".".join([function.__module__, function.__qualname__])
            config = params
            config["py>"] = command
        else:
            if isinstance(function, dict):
                config = function.copy()
                config.update(params)
            else:
                config = params
        try:
            json.dumps(config)
        except Exception as error:
            raise TypeError("Parameters must be serializable using JSON: %s" % str(error))
        self.subtask_config["+subtask" + str(self.subtask_index)] = config
        self.subtask_index += 1

digdag_mod.env = Env(digdag_env_mod)
import digdag

# add the archive path to import path
sys.path.append(os.path.abspath(os.getcwd()))

def digdag_inspect_command(command):
    # package.name.Class.method
    fragments = command.split(".")
    method_name = fragments.pop()
    class_type = None
    callable_type = None
    try:
        mod = __import__(".".join(fragments), fromlist=[method_name])
        try:
            callable_type = getattr(mod, method_name)
        except AttributeError as error:
            raise AttributeError("Module '%s' has no attribute '%s'" % (".".join(fragments), method_name))
    except ImportError as error:
        class_name = fragments.pop()
        mod = __import__(".".join(fragments), fromlist=[class_name])
        try:
            class_type = getattr(mod, class_name)
        except AttributeError as error:
            raise AttributeError("Module '%s' has no attribute '%s'" % (".".join(fragments), method_name))

    if type(callable_type) == type:
        class_type = callable_type
        method_name = "run"

    if class_type is not None:
        return (class_type, method_name)
    else:
        return (callable_type, None)

def digdag_inspect_arguments(callable_type, exclude_self, params):
    if callable_type == object.__init__:
        # object.__init__ accepts *varargs and **keywords but it throws exception
        return {}
    if hasattr(inspect, 'getfullargspec'): # Python3
        spec = inspect.getfullargspec(callable_type)
        keywords_ = spec.varkw
    else: # Python 2
        spec = inspect.getargspec(callable_type)
        keywords_ = spec.keywords

    args = {}
    for idx, key in enumerate(spec.args):
        if exclude_self and idx == 0:
            continue
        if key in params:
            args[key] = params[key]
        else:
            if spec.defaults is None or idx < len(spec.args) - len(spec.defaults):
                # this keyword is required but not in params. raising an error.
                if hasattr(callable_type, '__qualname__'):
                    # Python 3
                    name = callable_type.__qualname__
                elif hasattr(callable_type, 'im_class'):
                    # Python 2
                    name = "%s.%s" % (callable_type.im_class.__name__, callable_type.__name__)
                else:
                    name = callable_type.__name__
                raise TypeError("Method '%s' requires parameter '%s' but not set" % (name, key))
    if keywords_:
        # above code was only for validation
        return params
    else:
        return args

error = None
error_message = None
error_value = None
error_traceback = None
callable_type = None
method_name = None

try:
    callable_type, method_name = digdag_inspect_command(command)
    if method_name:
        init_args = digdag_inspect_arguments(callable_type.__init__, True, params)
        instance = callable_type(**init_args)

        method = getattr(instance, method_name)
        method_args = digdag_inspect_arguments(method, True, params)
        result = method(**method_args)
    else:
        args = digdag_inspect_arguments(callable_type, False, params)
        result = callable_type(**args)
except SystemExit as e:
    # SystemExit only shows an exit code and it is not kind to users. So this block creates a specific error message.
    # This error will happen if called python module name and method name are equal to those of the standard library module. (e.g. tokenize.main)
    error = Exception("Python command %s terminated with exit code %d" % (command, e.code), "Possible cause: program intentionally/accidentally finished, or module name conflicted with the standard library")
    error_type, error_value, _tb = sys.exc_info()
    error_message = "%s %s" % (error.args[0], error.args[1])
    error_traceback = traceback.format_exception(error_type, error_value, _tb)
except Exception as e:
    error = e
    error_type, error_value, _tb = sys.exc_info()
    error_message = str(error_value)
    error_traceback = traceback.format_exception(error_type, error_value, _tb)

out = {
    'subtask_config': digdag_env.subtask_config,
    'export_params': digdag_env.export_params,
    'store_params': digdag_env.store_params,
    #'state_params': digdag_env.state_params,  # only for retrying
}

if error:
    out['error'] = {
        'class': error_value.__class__.__name__,
        'message': error_message,
        'backtrace': error_traceback
    }

with open(out_file, 'w') as f:
    json.dump(out, f)

if error:
    raise error
